### PR TITLE
Add event support, use for disconnection and inputs

### PR DESCRIPTION
**Pull request by Manishearth**
Allow the device to send events back to script
r? @asajeffrey

### CODE REVIEW

**Comment by Manishearth** [comment]
Okay, session ending is a total mess. We need:

 - The ability for JS to request the session to end, and get an event when the session is fully shut down
 - The ability for the device to decide it needs to end, and send the same event.
 - The ability for mock APIs to ask the device to shut down and notice when it's done

Currently, there are a bunch of things preventing this from happening:
 - Mock events are only processed if there is a rAF
 - The device has no way to end the loop since it can't access it
 - There's no way for the headless device to "return" its receiver to the headless device

I think this will need a bunch of changes to work right.

I think we should just land this PR and https://github.com/servo/servo/pull/23814 first as it improves some of the tests and makes them less flaky, then overhaul how we do mocking. The changes we need:

 - Don't have multiple mock discovery objects, just have a single one
 - Provide session objects with "quitter"s that loop back to their own event loop
 - route mock events through the main device event loop, perhaps as a core part of the API (`Session.send_mock()`)

This will change a lot of things, thus the preference for landing the progress we have so far as-is to deal with intermittents.

File: webxr-api/device.rs [hunk 1]
```diff
@@ -57,4 +58,13 @@ pub trait Device: 'static {
...
     fn initial_inputs(&self) -> Vec<InputSource>;
+
+    /// Sets the event handling callback
+    fn set_event_callback(&mut self, callback: Box<dyn EventCallback>);
+
+    /// Whether the device is still connected
+    fn connected(&mut self) -> bool;
+
+    /// Quit the session
+    fn quit(&mut self);
```
Comment (asajeffrey) [inline webxr-api/device.rs]: Can we just make this the `Drop` code for the device?

File: webxr-api/device.rs [hunk 2]
```diff
@@ -57,4 +58,13 @@ pub trait Device: 'static {
     /// Inputs registered with the device on initialization. More may be added, which
     /// should be communicated through a yet-undecided event mechanism
     fn initial_inputs(&self) -> Vec<InputSource>;
+
+    /// Sets the event handling callback
+    fn set_event_callback(&mut self, callback: Box<dyn EventCallback>);
+
+    /// Whether the device is still connected
+    fn connected(&mut self) -> bool;
```
Comment (asajeffrey) [inline webxr-api/device.rs]: Should this be an event rather than a flag that is polled?

File: webxr-api/events.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+#[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "ipc", derive(serde::Serialize, serde::Deserialize))]
+pub enum Event {
+    /// Input source connected
+    AddInput(InputSource),
+    /// Input source disconnected
+    RemoveInput(InputId),
+    /// Session ended by device
+    SessionEnd,
```
Comment (asajeffrey) [inline webxr-api/events.rs]: Should there be an event for the device being disconnected too? Blur / focus events?

File: webxr/glwindow/mod.rs [hunk 4]
```diff
@@ -151,6 +156,19 @@ impl Device for GlWindowDevice {
...
+        self.events.upgrade(callback)
+    }
+
+    fn connected(&mut self) -> bool {
+        self.connected
+    }
+
+    fn quit(&mut self) {
+        self.connected = false;
+        self.events.callback(Event::SessionEnd);
```
Comment (asajeffrey) [inline webxr/glwindow/mod.rs]: Only send this event if the device is currently connected?

File: webxr/headless/mod.rs [hunk 5]
```diff
@@ -39,12 +46,22 @@ struct HeadlessDiscovery {
...
+
 struct HeadlessDevice {
     gl: Rc<dyn Gl>,
     floor_transform: TypedRigidTransform3D<f32, Native, Floor>,
     viewer_origin: TypedRigidTransform3D<f32, Viewer, Native>,
     views: Views,
     receiver: Receiver<MockDeviceMsg>,
+    events: EventBuffer,
+    inputs: Vec<InputInfo>,
+    disconnect_callbacks: Vec<Sender<()>>,
```
Comment (asajeffrey) [inline webxr/headless/mod.rs]: Why is disconnect treated differently from the other events?

File: webxr/headless/mod.rs [hunk 6]
```diff
@@ -114,6 +141,26 @@ impl Device for HeadlessDevice {
...
+        self.events.upgrade(callback)
+    }
+
+    fn connected(&mut self) -> bool {
+        if self.connected {
+            true
+        } else {
+            for callback in self.disconnect_callbacks.drain(..) {
+                let _ = callback.send(());
+            }
```
Comment (asajeffrey) [inline webxr/headless/mod.rs]: Calling this method has a side-effect? I wasn't expecting that!

File: webxr-api/device.rs (hunk 1 above)
Comment (Manishearth) [reply webxr-api/device.rs]: Then XRSession would have to store an `Option<Session>` for no good reason.

File: webxr-api/device.rs (hunk 2 above)
Comment (Manishearth) [reply webxr-api/device.rs]: This is the communication issue I need to fix

File: webxr-api/events.rs (hunk 3 above)
Comment (Manishearth) [reply webxr-api/events.rs]: Same thing as SessionEnd from the POV of the device IMO.

Blur/Focus yes, I just haven't gotten to it yet.

File: webxr/headless/mod.rs (hunk 5 above)
Comment (Manishearth) [reply webxr/headless/mod.rs]: The mock API requires we signal disconnection to have succeeded.

File: webxr/headless/mod.rs (hunk 6 above)
Comment (Manishearth) [reply webxr/headless/mod.rs]: Yeah, this isn't ideal, I want to change this in a followup.

File: webxr-api/events.rs (hunk 3 above)
Comment (Manishearth) [reply webxr-api/events.rs]: We should probably have VisibilityStateChange instead

**Pull request merged**